packages/next-upgrade/index.ts

```diff
@@ -0,0 +1,223 @@
+  if (nextPackageJson['latest'].version && nextPackageJson['rc'].version) {
+    showRc =
+      compareVersions(
+        nextPackageJson['rc'].version,
+        nextPackageJson['latest'].version
+      ) === 1
+  }
+
+  const choices = [
+    {
+      title: 'Canary',
+      value: 'canary',
+      description: `Experimental version with latest features (${nextPackageJson['canary'].version})`,
+    },
+  ]
+  if (showRc) {
+    choices.push({
+      title: 'Latest Release Candidate',
+      value: 'rc',
+      description: `Pre-release version for final testing (${nextPackageJson['rc'].version})`,
+    })
+  }
+  choices.push({
+    title: 'Latest Stable',
+    value: 'latest',
+    description: `Production-ready release (${nextPackageJson['latest'].version})`,
+  })
+
+  const shortcutVersion = process.argv[2]?.replace('@', '')
+  if (['canary', 'rc', 'latest'].includes(shortcutVersion)) {
+    // Shortcut
+    targetVersionSpecifier = shortcutVersion as VersionSpecifier
+  } else {
+    const initialVersionSpecifierIdx = await processCurrentVersion()
+
+    const response: Response = await prompts({
+      type: 'select',
+      name: 'version',
+      message: 'What Next.js version do you want to upgrade to?',
+      choices: choices,
+      initial: initialVersionSpecifierIdx,
+    })
+
+    if (!response.version) {
+      return
+    }
+
+    targetVersionSpecifier = response.version
+  }
+
+  // TODO(LichuAcu): support Turborepo
+  const appPackageJsonPath = path.resolve(process.cwd(), 'package.json')
+  let appPackageJson = JSON.parse(fs.readFileSync(appPackageJsonPath, 'utf8'))
+
+  const targetNextPackageJson = nextPackageJson[targetVersionSpecifier]
+
+  if (targetNextPackageJson.version) {
+    if (compareVersions(targetNextPackageJson.version, '15.0.0-canary') >= 0) {
+      await suggestTurbopack(appPackageJson)
+    }
+  }
+
+  fs.writeFileSync(appPackageJsonPath, JSON.stringify(appPackageJson, null, 2))
+
+  // TODO(LichuAcu): keep the original styling of the file
+  console.log('\nUpdating depencies...\n')
+
+  let packageManager: PackageManager = getPackageManager(appPackageJson)
+  const dependencies = `react@${targetNextPackageJson.peerDependencies['react']} react-dom@${targetNextPackageJson.peerDependencies['react-dom']} next@${targetVersionSpecifier}`
+
+  let updateCommand
+  switch (packageManager) {
+    case 'pnpm':
+      updateCommand = `pnpm update ${dependencies}`
+      break
+    case 'npm':
+      updateCommand = `npm install ${dependencies}`
+      break
+    case 'yarn':
+      updateCommand = `yarn add ${dependencies}`
+      break
+    default:
+      updateCommand = `pnpm install next@${targetVersionSpecifier}`
+      break
+  }
+
+  execSync(updateCommand, {
+    stdio: 'inherit',
+  })
+
+  appPackageJson = JSON.parse(fs.readFileSync(appPackageJsonPath, 'utf8'))
+  appPackageJson.dependencies['next'] = targetVersionSpecifier
+  fs.writeFileSync(appPackageJsonPath, JSON.stringify(appPackageJson, null, 2))
```

Comment:
Shouldn't this already happen as part of `updateCommand`? Do we really need to modify `package.json` ourselves as well?

Reply:
I'm doing this so that package.json ends up, for example, with `"next": "canary"` instead of `"next": "15.0.0-canary.156"`, so the user doesn't get stuck in a specific version Next.js version. When running `pnpm update next@canary`, pnpm will unfortunately write the specific version it used to the package.json.

I don't know how big of an issue it is if the package.json ends up with `"next": "15.0.0-canary.156"`. Not having to manually modify package.json is definitely good.